crates/bevy_pbr/src/light.rs

```diff
@@ -568,19 +784,135 @@ pub(crate) fn assign_lights_to_clusters(
 
     let light_count = lights.len();
     let mut global_lights_set = HashSet::with_capacity(light_count);
-    for (view_entity, view_transform, camera, frustum, mut clusters) in views.iter_mut() {
+    for (view_entity, view_transform, camera, frustum, config, mut clusters) in views.iter_mut() {
+        if matches!(config, ClusterConfig::None) {
+            commands.entity(view_entity).remove::<VisiblePointLights>();
+            continue;
+        }
+
         let view_transform = view_transform.compute_matrix();
         let inverse_view_transform = view_transform.inverse();
-        let cluster_count = clusters.aabbs.len();
         let is_orthographic = camera.projection_matrix.w_axis.w == 1.0;
+
+        let screen_size_u32 = camera.target.get_physical_size(&windows, &images);
+        if screen_size_u32.is_none() {
+            continue;
+        }
+        let screen_size_u32 = screen_size_u32.unwrap();
+        let mut cluster_dimensions = config.dimensions_for_screen_size(screen_size_u32);
+
+        let far_z = match config.far_z_mode() {
+            ClusterFarZMode::CameraFarPlane => camera.far,
+            ClusterFarZMode::MaxLightRange => lights
+                .iter()
+                .map(|light| {
+                    (inverse_view_transform * light.translation.extend(1.0)).z * -1.0 + light.range
+                })
+                .reduce(f32::max)
+                .unwrap_or(0.0),
+            ClusterFarZMode::Constant(far) => far,
+        };
+        let first_slice_depth = match cluster_dimensions.z {
+            1 => config.first_slice_depth().max(far_z),
+            _ => config.first_slice_depth(),
+        };
+        let far_z = far_z.max(first_slice_depth);
+
         let cluster_factors = calculate_cluster_factors(
-            // NOTE: Using the special cluster near value
-            clusters.near,
-            camera.far,
-            clusters.axis_slices.z as f32,
+            first_slice_depth,
+            far_z,
+            cluster_dimensions.z as f32,
             is_orthographic,
         );
 
+        let max_indices = ViewClusterBindings::MAX_INDICES;
+
+        if max_indices
+            < (cluster_dimensions.x * cluster_dimensions.y * cluster_dimensions.z) as usize
+                * light_count
```

Comment:
So if the maximum indices is >= the number of indices that could possibly be used for the cluster configuration, we wouldn't do any estimation and reconfiguration? But if that is the case, we could probably get a better configuration that made use of more clusters. My feeling is that this estimation and adjustment process should depend rather on the ClusterConfig. Perhaps just another bool to determine if the dynamic approach should be used?

Reply:
my thought here was mainly to avoid running the test when we're using unbounded storage buffers.

that said, we could have a situation where the user knows max_indices won't be exceeded due to the specific setup of their lights/clustering, so i agree a toggle makes sense, will add.